### PR TITLE
Heurisch

### PR DESCRIPTION
**Pull request by vishalg2235**
Fixes https://github.com/sympy/sympy/issues/13723

### CODE REVIEW

**Comment by vishalg2235** [comment]
@jksuom Please review. Thanks

**Comment by vishalg2235** [comment]
@jksuom I still didn't understand how heurisch.py work (mathematically or algorithm wise). Can you suggest me where I can study those algorithms ?

**Comment by jksuom** [comment]
It not easy to find freely accessible sources for the algorithm. It is often called "parallel Risch algorithm" or "Risch-Norman algorithm". You can try to find sources in the net. A very brief description is this: It follows from Liouville's Theorem that the antiderivative is a linear combination of certain terms that can be derived from the integrand. The algorithm then uses the method of unknown coefficients to find the solution, if any.

**Comment by vishalg2235** [comment]
@jksuom I was trying to fix https://github.com/sympy/sympy/issues/13749 with 'new'(commited one) heurisch.py then it's integral came in terms in terms of `atan`, but while evaluating it from limits here [https://github.com/sympy/sympy/blob/master/sympy/integrals/integrals.py#L579](url) it didn't check whether it is `piecewise` or not. I even tried  `integrate(1/(cos(t) + 2), (t,0,2*pi),conds = 'piecewise')` but still it goes to https://github.com/sympy/sympy/blob/master/sympy/integrals/integrals.py#L579 . How can we make it piecewise limits ?

**Comment by jksuom** [comment]
What do you mean by "piecewise limits"?

**Comment by vishalg2235** [comment]
Sorry piecewise integration. The limits must be broke down into 0->pi and pi->2*pi .

**Comment by jksuom** [comment]
The evaluation of definite integrals of periodic functions is an issue that cannot expected to be solvable by simply rewriting log terms as atans. That is the subject of a different pull request (and probably not easy to solve).

**Comment by vishalg2235** [comment]
ok !

**Comment by asmeurer** [comment]
I would add more tests, including tests for logs that don't fully match to make sure that no terms are lost.

**Comment by vishalg2235** [comment]
@jksuom @asmeurer can I work on evaluation of definite integrals of periodic functions ?

**Comment by vishalg2235** [comment]
@jksuom @asmeurer one thing I haven't noticed yet that for `integrate(1/(cos(t) + 2), (t,0,2*pi))` the result:  `2*sqrt(3)*atan(sqrt(3)*tan(t/2)/3)/3` is wrong as in 0 to pi this function is +ve and in pi to 2*pi its -ve (and symmetrically inverse).

**Comment by jksuom** [comment]
It is only the main branch of `atan` that is negative. One has to choose another branch taking positive values in the interval `[pi, 2*pi]` and having no discontinuity at `pi`. (Multivalued antiderivatives are the general issue with definite integrals of periodic functions.)

**Comment by vishalg2235** [comment]
Ok so `2*sqrt(3)*atan(sqrt(3)*tan(t/2)/3)/3` should have came within `modulus` sign.
@jksuom I think this(definite integrals of periodic functions) is a big issue which has to be fixed. is there anyone working on this topic? Can I help or try to fix it?

**Comment by vishalg2235** [comment]
Fixes #9569 .

**Comment by jksuom** [comment]
> is there anyone working on this topic?

No one as far as I know.

> Can I help or try to fix it?

You are welcome to do that. (But that may be a big task.)

**Comment by vishalg2235** [comment]
@jksuom Yup! I would like to do it and will try my best. 
So my task will be to find definite integral of periodic functions.
We can include a new argument "periodic" in integrate and if it's `True` then implement those rules(code) of periodic definite integrals. can you give me some reference how to start implementing it?

**Comment by jksuom** [comment]
There is a link in [this comment](https://github.com/sympy/sympy/issues/13749#issuecomment-352168314).

**Comment by vishalg2235** [comment]
It is failing at tests : [here](https://github.com/sympy/sympy/blob/master/sympy/integrals/tests/test_heurisch.py#L151) and [here](https://github.com/sympy/sympy/blob/master/sympy/integrals/tests/test_heurisch.py#L205) . 
But their integrals are coming as follows: 
```
>>> y = symbols('y')
>>> integrate(1/(x**2 + y), x)
-sqrt(-1/y)*log(x - y*sqrt(-1/y))/2 + sqrt(-1/y)*log(x + y*sqrt(-1/y))/2
>>> integrate(1/(x * (1 + log(x)**2)), x)
RootSum(4*_z**2 + 1, Lambda(_i, _i*log(2*_i + log(x))))
```
In first test case integral is coming `real`, so how it can be asserted to `complex` expression.
In second test case I don't understand where error is coming because in `master` also same integral comes. And how `RootSum` works?

**Comment by jksuom** [comment]
In the first test, the complex expressions come from the factorization `x**2 + y = (x + I*sqrt(y))*(x - I*sqrt(y)) `. The test should be updated, probably as follows:

    assert heurisch_wrapper(1/(x**2 + y), x) == atan(x/sqrt(y))/sqrt(y)

Also the second test should become

    assert heurisch(1/(x*(1 + log(x)**2)), x) == atan(log(x))

The `RootSum` in the result of `integrate` is coming from another integrator (`risch_integrate`). For a polynomial `p` and a function `f`, `RootSum(p, f)` denotes the sum of `f(x)` over all roots `x` of `p`.

**Comment by vishalg2235** [comment]
@jksuom why these errors is coming?
First one is integrated from `ratint` and second one from `risch_integrate`. Neither I made changes in `ratint` nor in `risch_integrate` so why they are occurring ?

**Comment by vishalg2235** [comment]
Ping @jksuom .

**Comment by jksuom** [comment]
This is what I get with this branch:
```
In [3]: heurisch(1/(x * (1 + log(x)**2)), x)
Out[3]: atan(log(x))
```
I think this is the expected result. Only the failing test should be updated accordingly as I have indicated above. The other test should also be updated.

**Comment by vishalg2235** [comment]
@jksuom please review!

File: sympy/integrals/heurisch.py [hunk 1]
```diff
@@ -594,21 +596,49 @@ def _integrate(field=None):
...
+            x, y = pairs.pop()
+            if (x,-y) in pairs:
+                pairs.remove((x,-y))
+                # Choosing b with no minus sign
+                if y.could_extract_minus_sign():
+                    y = -y
+                irreducibles.add(x*x + y*y)
+                atans.add(atan(x/y))
+            else:
+                irreducibles.add(x+I*y)
```
Comment (jksuom) [inline sympy/integrals/heurisch.py]: Space on either side of ``+`.

**Comment by jksuom** [comment]
I think this looks good. New tests should be added. Some of those can be created by taking the derivative of suitable functions. There are also open issues of the type "value of integral is wrong" (often labelled 'integrals.definite').

**Comment by vishalg2235** [comment]
https://github.com/sympy/sympy/issues/13749 this will be solved once this PR(gives integral `2*sqrt(3)*atan(sqrt(3)*tan(x/2)/3)/3` terms) and [this](https://github.com/sympy/sympy/pull/13808) (adds floor term as there is `atan` with args linear function of tan) both will be fixed.

**Comment by jksuom** [comment]
Yes, that is what I expect.

**Comment by vishalg2235** [comment]
```
>>> diff(atan(sin(x)**2))
2*sin(x)*cos(x)/(sin(x)**4 + 1)
>>> heurisch(2*sin(x)*cos(x)/(sin(x)**4 + 1),x)
atan(sqrt(2)*sin(x) - 1) - atan(sqrt(2)*sin(x) + 1)
```
Here integral could be simplified more to `atan(-1/cos(x)**2)` or `pi/2 + atan(cos(x)**2)`, after combining both `atan` terms and then using atan properties.

Is it okay if I add test: 
```
assert heurisch(2*sin(x)*cos(x)/(sin(x)**4 + 1), x) == atan(sqrt(2)*sin(x) - 1) - atan(sqrt(2)*sin(x) + 1)
```

**Comment by jksuom** [comment]
That test is probably OK.

**Comment by jksuom** [comment]
This could be one of the tests: `heurisch((2*sin(x)*cos(x) + cos(x))/(1 + sin(x)**2), x)`.

**Comment by vishalg2235** [comment]
@jksuom ErrorLog says maximum time limit exceeded. Please have a look on the checks!

**Comment by jksuom** [comment]
I have restarted the failing test: `curl: (7) Failed to connect to s3.amazonaws.com port 443: Connection timed out`

**Comment by vishalg2235** [comment]
Should I add more tests or its fine ?

**Comment by jksuom** [comment]
I would like to have more tests performed though not necessarily added to the test files since they tend to take a lot of time. Are there open issues that would provide integrands?

**Comment by vishalg2235** [comment]
I have gone through issues labelled and `integrals.definite` but only found one test `integrate(1/(cos(x)+2))` now I'am searching it in `integrals` label.

**Comment by jksuom** [comment]
@vishalg2235 It seems to me that #14064 would provide an example that could be added. You could also take an example that would create both log and atan terms such as
`integrate((1 + sin(x))/(2 + cos(x)), x)`

**Comment by vishalg2235** [comment]
@jksuom please review! Error says `global name 'Ne' is not defined` but in heurisch.py it is defined.

File: sympy/integrals/tests/test_heurisch.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 from sympy import Rational, sqrt, symbols, sin, exp, log, sinh, cosh, cos, pi, \
-    I, erf, tan, asin, asinh, acos, Function, Derivative, diff, simplify, \
-    LambertW, Ne, Piecewise, Symbol, Add, ratsimp, Integral, Sum, \
-    besselj, besselk, bessely, jn
+    I, erf, tan, asin, asinh, acos, atan, Function, Derivative, diff, simplify, \
+    LambertW, Eq, Piecewise, Symbol, Add, ratsimp, Integral, Sum, \
```
Comment (jksuom) [inline sympy/integrals/tests/test_heurisch.py]: `Ne` should also be imported.

**Comment by vishalg2235** [comment]
Now one ValueError is raised from test_perm_group.py that "there were repeated elements". How to fix it
?

**Comment by jksuom** [comment]
There were several transmission errors (5, in fact). I have restarted the jobs.

**Comment by jksuom** [comment]
The results look good and this can be merged. Thank you @vishalg2235.

**Comment by vishalg2235** [comment]
Thank you @jksuom for your constant help :)

**Pull request merged**